suite/chaos/reorg_confirmer.go

```diff
@@ -0,0 +1,254 @@
+const (
+	TXNodeLabel    = "geth-reorg-ethereum-geth"
+	MinerNodeLabel = "geth-reorg-ethereum-miner-node"
+)
+
+// ReorgConfirmer reorg stats collecting struct
+type ReorgConfirmer struct {
+	env     *environment.Environment
+	clients []client.BlockchainClient
+	// number of blocks we want to be reorged on node zero (default client)
+	reorgDepth int
+	// blocks for that all nodes sent equal hashes, amount of blocks to wait
+	blockConsensusThreshold int
+	numberOfNodes           int
+	currentBlockConsensus   int
+	forkBlockNumber         int64
+	currentAltBlocks        int
+	currentVerifiedBlocks   int
+	NetworkStep             ReorgStep
+	altBlockNumbers         []int64
+	blocksByNode            map[int]map[int64]client.NodeBlock
+	blockHashes             map[int64][]common.Hash
+	chaosExperimentName     string
+	mutex                   sync.Mutex
+	ctx                     context.Context
+	cancel                  context.CancelFunc
+	doneChan                chan struct{}
+	done                    bool
+}
+
+// NewReorgConfirmer creates a type that can create reorg chaos and confirm reorg has happened
+func NewReorgConfirmer(
+	clients []client.BlockchainClient,
+	env *environment.Environment,
+	reorgDepth int,
+	blockConsensusThreshold int,
+	timeout time.Duration,
+) (*ReorgConfirmer, error) {
+	if len(clients) == 1 {
+		return nil, errors.New("only one node within the blockchain client detected, cannot reorg")
+	}
+	ctx, ctxCancel := context.WithTimeout(context.Background(), timeout)
+	rc := &ReorgConfirmer{
+		env:                     env,
+		clients:                 clients,
+		reorgDepth:              reorgDepth,
+		blockConsensusThreshold: blockConsensusThreshold,
+		numberOfNodes:           len(clients),
+		doneChan:                make(chan struct{}),
+		altBlockNumbers:         make([]int64, 0),
+		blockHashes:             map[int64][]common.Hash{},
+		blocksByNode:            map[int]map[int64]client.NodeBlock{},
+		mutex:                   sync.Mutex{},
+		ctx:                     ctx,
+		cancel:                  ctxCancel,
+	}
+	return rc, nil
+}
+
+// ReceiveBlock receives block marked by node that mined it,
+// forks the network and record all alternative block numbers on node 0
+func (rc *ReorgConfirmer) ReceiveBlock(blk client.NodeBlock) error {
+	rc.mutex.Lock()
+	defer rc.mutex.Unlock()
+	if blk.Block == nil || rc.done {
+		return nil
+	}
+	rc.appendBlockHeader(blk)
+	switch rc.NetworkStep {
+	case InitConsensus:
+		if rc.hasNetworkFormedConsensus(blk) {
+			rc.NetworkStep = NetworkFork
+		}
+	case NetworkFork:
+		if err := rc.forkNetwork(blk); err != nil {
+			return err
+		}
+	case AltBlocks:
+		if err := rc.awaitAltBlocks(blk); err != nil {
+			return err
+		}
+	case Consensus:
+		if rc.hasNetworkFormedConsensus(blk) {
+			rc.done = true
+			rc.doneChan <- struct{}{}
+		}
+	}
+	return nil
+}
+
+// Verify verifies that all blocks are replaced by reorg
+func (rc *ReorgConfirmer) Verify() error {
```

Comment:
Maybe worth renaming this something a bit more explicit, like `VerifyReorgComplete`?